Federation service controller

mfanjie opened the pull request:
``` release-note
Federation service controller is one key component of federation controller manager, it watches federation service, creates/updates services to all registered clusters, and update DNS records to global DNS server.
```

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

Commit:
add federation service controller

mfanjie commented:
@quinton-hoole I created this PR based on #26020, when you finish code review on the wip #26034, I will update all changes on both branches, so we can be ready for code merge.

quinton-hoole commented:
ok to test

quinton-hoole commented:
Thanks @mfanjie !

quinton-hoole commented:
Fixes #23848

quinton-hoole commented:
Still a few minor govet and flags issues to resolve, then LGTM.

```
FAILED   ./hack/../hack/verify-gofmt.sh	9s
Verifying ./hack/../hack/verify-govet.sh
federation/pkg/federation-controller/service/endpoint_helper.go:38: range variable clusterCache captured by func literal
federation/pkg/federation-controller/service/endpoint_helper.go:43: range variable clusterCache captured by func literal
federation/pkg/federation-controller/service/endpoint_helper.go:44: range variable clusterName captured by func literal
federation/pkg/federation-controller/service/endpoint_helper.go:44: range variable clusterCache captured by func literal
federation/pkg/federation-controller/service/service_helper.go:41: range variable clusterCache captured by func literal
federation/pkg/federation-controller/service/service_helper.go:46: range variable clusterCache captured by func literal
federation/pkg/federation-controller/service/service_helper.go:47: range variable clusterName captured by func literal
federation/pkg/federation-controller/service/service_helper.go:47: range variable clusterCache captured by func literal
federation/pkg/federation-controller/service/servicecontroller.go:291: range variable clusterName captured by func literal
federation/pkg/federation-controller/service/servicecontroller.go:291: range variable cluster captured by func literal
federation/pkg/federation-controller/service/endpoint_helper_test.go:32: k8s.io/kubernetes/pkg/api/v1.EndpointAddress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:32: k8s.io/kubernetes/pkg/api/v1.LoadBalancerIngress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:56: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:64: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:66: k8s.io/kubernetes/pkg/api/v1.LoadBalancerIngress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:69: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:81: k8s.io/kubernetes/pkg/api/v1.LoadBalancerIngress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:84: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:92: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:97: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:129: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:131: k8s.io/kubernetes/pkg/api/v1.LoadBalancerIngress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:134: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:142: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:148: k8s.io/kubernetes/pkg/api/v1.ServiceStatus composite literal uses unkeyed fields
FAILED   ./hack/../hack/verify-govet.sh	89s
```

```
Verifying ./hack/../hack/verify-flags-underscore.py
Found flags in golang files not in the list of known flags. Please add these to hack/verify-flags/known-flags.txt
concurrent-service-syncs
FAILED   ./hack/../hack/verify-flags-underscore.py	5s
```

mfanjie commented:
resolved above problems and squashed commit.

nikhiljindal commented:
verify-govet is still failing. From logs:

```
Verifying ./hack/../hack/verify-govet.sh
federation/pkg/federation-controller/service/endpoint_helper_test.go:32: k8s.io/kubernetes/pkg/api/v1.EndpointAddress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:31: k8s.io/kubernetes/pkg/api/v1.LoadBalancerIngress composite literal uses unkeyed fields
FAILED   ./hack/../hack/verify-govet.sh
```

quinton-hoole commented:
@mfanjie govet is still not quite happy

```
Verifying ./hack/../hack/verify-govet.sh
federation/pkg/federation-controller/service/endpoint_helper_test.go:32: k8s.io/kubernetes/pkg/api/v1.EndpointAddress composite literal uses unkeyed fields
federation/pkg/federation-controller/service/service_helper_test.go:31: k8s.io/kubernetes/pkg/api/v1.LoadBalancerIngress composite literal uses unkeyed fields
FAILED   ./hack/../hack/verify-govet.sh	101s
```

quinton-hoole commented:
Snap :-)

mfanjie commented:
acked

mfanjie commented:
pushed again, I will monitor the e2e result.

quinton-hoole commented:
Looks good so far:
```
Verifying ./hack/../hack/verify-govet.sh
SUCCESS  ./hack/../hack/verify-govet.sh	91s
```

quinton-hoole commented:
Now only blocked on #26026, which is in turn blocked on #26105 

```
federation/pkg/dnsprovider/providers/google/clouddns/clouddns.go:27:2: cannot find package "google.golang.org/api/dns/v1" in any of:
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/api/dns/v1 (vendor tree)

```

mfanjie commented:
I fixed two minor bugs today, one test suite failed and it should be an test env problem?
```
ERROR: (gcloud.compute.copy-files) [/usr/bin/scp] exited with return code [1]. See https://cloud.google.com/compute/docs/troubleshooting#ssherrors for troubleshooting hints.

Test Suite Failed
!!! Error in /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce@2/kubernetes/hack/ginkgo-e2e.sh:95
```

quinton-hoole commented:
```
federation/pkg/dnsprovider/providers/google/clouddns/clouddns.go:27:2: cannot find package "google.golang.org/api/dns/v1" in any of:
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/api/dns/v1 (vendor tree)
	/usr/local/go/src/google.golang.org/api/dns/v1 (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/google.golang.org/api/dns/v1 (from $GOPATH)

```

This will be resolved once #26020 merges.

quinton-hoole commented:
@k8s-bot test this issue: #26129

mfanjie commented:
fixed several bugs found by integration test with cluster controller, now the service controller is much more stable than last version. I pushed the changes and squashed commit. 

I will hold any further changes until this PR being merged from now on.

nikhiljindal commented on federation/cmd/federation-controller-manager/app/controllermanager.go:
```diff
@@ -103,7 +107,17 @@ func Run(s *options.CMServer) error {
+	if err != nil {
+		glog.Fatalf("Cloud provider could not be initialized: %v", err)
+	}
+	scclientset := internalclientset.NewForConfigOrDie(restclient.AddUserAgent(restClientCfg, "federation-service-controller"))
```
Use service.UserAgent (the const)

nikhiljindal commented on federation/pkg/federation-controller/service/cluster_helper.go:
```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+// addToClusterSet insert the new cluster to clusterSet and create a corresponding
```
insert -> inserts, create -> creates

nikhiljindal commented on federation/pkg/federation-controller/service/endpoint_helper.go:
```diff
@@ -0,0 +1,165 @@
+// Whenever there is change on endpoint, the federation service should be updated
+// key is the namespaced name of endpoint
+func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache *clusterCache, serviceCache *serviceCache, fedClient federationclientset.Interface) error {
+	cachedService, ok := serviceCache.get(key)
```
should this be cachedEndpoint, ok := endpointCache.get(key)?

mfanjie replied:
Here we check if the service cache exists, as the watcher catches all endpoints from clusters, if the endpoint is created by k8s not federation, there will no serviceCache, and we should simply skip that.
serviceCache->fedServiceMap:[clustername]*cachedService->endpointMap[string]int

nikhiljindal commented on federation/pkg/federation-controller/service/endpoint_helper.go:
```diff
@@ -0,0 +1,165 @@
+		// here we filtered all non-federation services
+		return nil
+	}
+	// obj holds the latest service info from apiserver
```
Whats obj here?

nikhiljindal commented on federation/pkg/federation-controller/service/endpoint_helper.go:
```diff
@@ -0,0 +1,165 @@
+		glog.Infof("can not get endpoint %v for cluster %s from endpointStore", key, clusterName)
+		err = cc.processEndpointDeletion(cachedService, clusterName)
+	}
+	if exists {
```
Make this else.

nikhiljindal commented on federation/pkg/federation-controller/service/endpoint_helper.go:
```diff
@@ -0,0 +1,165 @@
+	return nil
+}
+
+// obj could be an *api.Service, or a DeletionFinalStateUnknown marker item.
```
should this be api.Endpoint instead of api.Service?

nikhiljindal commented on federation/pkg/federation-controller/service/service_helper.go:
```diff
@@ -0,0 +1,254 @@
+	// obj holds the latest service info from apiserver, return if there is no federation cache for the service
+	cachedService, ok := serviceCache.get(key)
+	if !ok {
+		// here we filtered all non-federation services
```
Why is this func called on non-federation services?

mfanjie replied:
if serviceCache does not exists, that means the service is not created by federation, we should skip it

nikhiljindal commented on federation/pkg/federation-controller/service/service_helper.go:
```diff
@@ -0,0 +1,254 @@
+			needUpdate = cc.processServiceDeletion(cachedService, clusterName)
+		}
+	}
+	if !exists {
```
same. make this else for the previous if

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+
+	doNotRetry = time.Duration(0)
+
+	UserAgentName = "Federation-Service-Controller"
```
Its federation-service-controller (all small) in controller manager. Should be the same at both places

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+
+type cachedService struct {
+
+	//release_1_3 client returns v1 service
```
Is this comment relevant here?

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+	appliedState *api.Service
+	// cluster endpoint map hold subset info from kubernetes clusters
+	// key clusterName
+	// value is a flag that if there is ready addresses
```
Is this comment correct?

mfanjie replied:
Yes, we do not care about the detail of endpoint, we only care about if there is ready address for that endpoint, so in the map, we only set flag.

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+type ServiceController struct {
+	dns              dnsprovider.Interface
+	federationClient federationclientset.Interface
+	// each federation should be configured with a single zone (e.g. "mycompany.com")
```
Why?

mfanjie replied:
I did not handle any logic related to zones, this part relies on Quinton. removed comment.

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+	clusterController *framework.Controller
+	eventBroadcaster  record.EventBroadcaster
+	eventRecorder     record.EventRecorder
+	//clusterLister     federationcache.StoreToClusterLister
```
delete if not required

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+func New(federationClient federationclientset.Interface, dns dnsprovider.Interface) *ServiceController {
+	broadcaster := record.NewBroadcaster()
+	// federationClient event is not supported yet?
+	//broadcaster.StartRecordingToSink(&unversioned_core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
```
Yes. no federation events for now

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+	broadcaster := record.NewBroadcaster()
+	// federationClient event is not supported yet?
+	//broadcaster.StartRecordingToSink(&unversioned_core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "federated-service-controller"})
```
federation-service-controller

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+	s.queue.Add(key)
+}
+
+// Run starts a background goroutine that watches for changes to Federated services
```
federation services

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+	return nil, !retryable
+}
+
+func (s *ServiceController) deleteClusterService(clusterName string, cachedService *cachedService, clientset *clientset.Clientset) (error, bool) {
```
What does the returned bool indicate here?

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+			}
+
+			if needUpdate {
+				// we only apply spec update for now
```
why "for now". When will we ever update the status?

If we ever update status then there is a chance that federation service controller and k8s service controller can end up overwriting each others status updates so we should prevent that if possible.

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+					glog.V(5).Infof("service %s/%s successfully updated to cluster %s", svc.Namespace, svc.Name, clusterName)
+					return nil
+				} else {
+					glog.V(4).Infof("failed to update %+v", err)
```
Just return err here?

mfanjie replied:
it's inside the loop, return error after retry.

nikhiljindal commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -0,0 +1,885 @@
+			}
+		}
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
```
nit: Some infof start with a capital letter, some dont :)
Will be good to be consistent.

nikhiljindal commented:
Still trying to understand the code. Have added a few comments.

nikhiljindal commented:
cc @kubernetes/sig-cluster-federation

mfanjie commented:
@nikhiljindal all your comments addressed.

quinton-hoole commented:
LGTM!  Needs to rebase against upstream/maser once #26020 merges (hopefully within the next few hours at most). Then ready for merge, once checks all pass.

quinton-hoole commented:
@mfanjie If you rebase this I should be able to merge it now.

mfanjie commented:
@quinton-hoole rebased and pushed

mfanjie commented:
failed for one checking, working on it.

In addition, I found some error on clustercontroller.go, even it is not related to this PR but it impacts the cluster status, so we have to manually update the cluster status to make service controller working properly, opened issue #26480 26480.

quinton-hoole commented:
Thanks @mfanjie.  As soon as the checks pass, I'll merge this.

quinton-hoole commented:
All checks have passed. Merge queue is blocked, so manually merging, as this is a v1.3 blocker, with several other blocker PR's waiting for it to merge.

mfanjie commented:
this is a sub task of #23653 , add linkage.

fd commented on federation/cmd/federation-controller-manager/app/options/options.go:
```diff
@@ -32,6 +32,14 @@ type ControllerManagerConfiguration struct {
+	// dnsProvider is the provider for dns services.
+	DnsProvider string `json:"dnsProvider"`
+	// dnsConfigFile is the path to the dns provider configuration file.
+	DnsConfigFile string `json:"ndsConfigFile"`
```
@quinton-hoole @mfanjie `ndsConfigFile` probably should be `dnsConfigFile`

(still exists on [master](https://github.com/mfanjie/kubernetes/blob/master/federation/cmd/federation-controller-manager/app/options/options.go#L38))

nikhiljindal replied:
You are right.
Being fixed in https://github.com/kubernetes/kubernetes/pull/27158

Though the json field name is not useful yet.
